velox/functions/sparksql/ConcatWsFunctions.cpp

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <utility>
+
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+// concat_ws(separator, string1, ..., stringN) → varchar
+// Returns the concatenation of string1, string2, ..., stringN connected by
+// separator.
+class ConcatWsFunction : public exec::VectorFunction {
+ public:
+  ConcatWsFunction(
+      const std::string& name,
+      const std::vector<exec::VectorFunctionArg>& inputArgs) {
+    auto numArgs = inputArgs.size();
+    VELOX_USER_CHECK(
+        numArgs >= 2,
+        "{} requires 2 arguments at least, but got {}",
+        name,
+        numArgs);
+
+    BaseVector* constantPattern = inputArgs[0].constantValue.get();
+    VELOX_USER_CHECK(
+        constantPattern != nullptr,
+        "{} requires first argument non-null",
+        name);
+
+    if (constantPattern->isNullAt(0)) {
+      nullDelim_ = true;
+      return;
+    }
+
+    delim_ =
+        constantPattern->as<ConstantVector<StringView>>()->valueAt(0).str();
+
+    // Save constant values to constantStrings_.
+    // Identify and combine consecutive constant inputs.
+    argMapping_.reserve(numArgs - 1);
+    constantStrings_.reserve(numArgs - 1);
+
+    for (auto i = 1; i < numArgs; ++i) {
+      argMapping_.push_back(i);
+
+      const auto& arg = inputArgs[i];
+      if (arg.constantValue) {
+        std::string value = arg.constantValue->as<ConstantVector<StringView>>()
+                                ->valueAt(0)
+                                .str();
+
+        column_index_t j = i + 1;
+        for (; j < inputArgs.size(); ++j) {
+          if (!inputArgs[j].constantValue) {
+            break;
+          }
+
+          value += delim_ +
+              inputArgs[j]
+                  .constantValue->as<ConstantVector<StringView>>()
+                  ->valueAt(0)
+                  .str();
+        }
+
+        constantStrings_.push_back(std::string(value.data(), value.size()));
```

Comment:
Can we write to the string raw buffer directly and avoid these intermediate strings? E.g., by using std::to_chars. This could help improve the performance.